### ISSUE TITLE
Requesting cache items concurrently throws NullReferenceException

### ISSUE THREAD

**Issue by dusda** [aspnet/Caching#270]
I am trying to run a few tasks at once to construct a common object, like so:

```
public async Task<SomeThing> GetAsync(string userId)
{
    var thing = new SomeThing(userId);

    var tasks = new[] { GetNameAsync(userId), GetEmailAsync(userId) };
    var res = await Task.WhenAll(tasks);

    profile.Name = res[0];
    profile.Email = res[1];

    return thing;
}
```

Here's an example of one of the tasks, they're all implemented the same way:

```
public async Task<string> GetEmailAsync(string userId)
{
    string email = await cache.GetStringAsync($"user:{userId}:email"));
    return email;
}
```

When I run this, it will *sometimes* throw a NullReferenceException. Here's the stack trace:

```
   at Microsoft.Extensions.Caching.Redis.RedisExtensions.<HashMemberGetAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Extensions.Caching.Redis.RedisCache.<GetAndRefreshAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Extensions.Caching.Redis.RedisCache.<GetAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Extensions.Caching.Distributed.DistributedCacheExtensions.<GetStringAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Rentluck.Stores.RedisUserProfileStore.<GetEmailAsync>d__7.MoveNext() in c\Projects\rentluck\src\Rentluck\Stores\IUserProfileStore.cs:line 75
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Rentluck.Stores.RedisUserProfileStore.<GetAsync>d__3.MoveNext() in c\Projects\rentluck\src\Rentluck\Stores\IUserProfileStore.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Rentluck.Services.UserProfileService.<GetUserProfileAsync>d__3.MoveNext() in c\Projects\rentluck\src\Rentluck\Services\UserProfileService.cs:line 27
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Rentluck.ProfileRefresher.Program.<>c__DisplayClass3_0.<<MainAsync>b__0>d.MoveNext() in c\Projects\rentluck\src\Rentluck.ProfileRefresher\Program.cs:line 39
```

If I rewrite the method to just get them synchronously it works fine every time.

FYI, I am using Microsoft's IDistributedCache for Redis, and StackExchange.Redis separately for some pub/sub at the same time, in a console application. Here's my NinectModule:

```
using Microsoft.Extensions.Caching.Distributed;
using Microsoft.Extensions.Caching.Redis;
using Microsoft.Extensions.Options;
using Ninject.Modules;
using StackExchange.Redis;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Rentluck.Modules
{
    public class RedisModule : NinjectModule
    {
        public override void Load()
        {
            //ms distributed cache
            Bind<IDistributedCache>().To<RedisCache>();
            Bind<IOptions<RedisCacheOptions>>().ToMethod((ctx) =>
                Options.Create(new RedisCacheOptions
                {
                    Configuration = ConfigurationManager.AppSettings["RedisConfiguration"],
                    InstanceName = "Rentluck"
                }));

            //pub/sub
            Bind<IPublisher>().To<RedisPublisher>();
            Bind<ISubscriber>().ToMethod((ctx) => Redis.Connection.GetSubscriber());
        }
    }
}
```

And here is how I'm managing the ConnectionMultiplexer for the pub/sub:

```
public class Redis
{
    static Lazy<ConfigurationOptions> options =
        new Lazy<ConfigurationOptions>(() => {
            var config = new ConfigurationOptions();
            config.EndPoints.Add(ConfigurationManager.AppSettings["RedisConfiguration"]);
            config.ClientName = "RentluckPubSub";
            config.ConnectTimeout = 30000;
            config.SyncTimeout = 30000;
            config.AbortOnConnectFail = false;
            return config;
        });

    static Lazy<ConnectionMultiplexer> con = new Lazy<ConnectionMultiplexer>(
        () => ConnectionMultiplexer.Connect(Options.Value));

    public static ConnectionMultiplexer Connection
    {
        get
        {
            return con.Value;
        }
    }

    public static Lazy<ConfigurationOptions> Options { get => options; set => options = value; }
}
```

**Comment by JunTaoLuo** [comment]
My initial guess is that this may be related to https://github.com/aspnet/Caching/issues/253. 

Before that was fixed, we may be opening multiple connections simultaneously. In these cases, by the time we retrieve the database from the given connection, that connection may have been lost and overwritten. The documentation doesn't specify what happens in this case but I suspect this can lead to some invalid states. I'll need to construct a repro to be certain that this is the cause for NREs though.

In the meantime, the fix for the aforementioned issue has already been made and back ported to LTS/FTS which should soon be available. It may be a good idea to wait and see if that solves the problem.

**Comment by JunTaoLuo** [comment]
#253 could also lead to a null ref where the connection is set but hasn't initialized the cache in Thread 1 and Thread 2 comes a long and sees the connection is set and tries to use the uninitialized cache. Again, this is fixed in LTS/FTS so it's worth seeing if the new packages fixes the problem.

**Comment by JunTaoLuo** [comment]
@dusda We have now released Microsoft.Extensions.Caching.Redis 1.1.1 and 1.0.2, can you give this another try?

**Comment by JunTaoLuo** [comment]
@dusda I believe the issue was fixed by the patches in FTS and LTS. Please re-open if you still see this after updating to the latest patches.

**Comment by hach-que** [comment]
@JunTaoLuo We're seeing this on Microsoft.Extensions.Caching.Redis 2.0.0, occurring very rarely, but the exact same stack trace and exception type as reported in this issue.

**Comment by JunTaoLuo** [comment]
@hach-que can you provide more detail on how you are using the redis cache? It would be great if you could provide a repro we can take a look at.

**Comment by hach-que** [comment]
We're using it to cache Google Cloud Datastore entities. We don't have a reliable repro (it happens maybe once on the build server every 25 or so builds, where each build runs a hundred tests against multiple entities each).

Right now we've worked around the issue in our retry layer by adding a catch for NullReferenceException. Our retry layer looks something like this:

```csharp
using Microsoft.Extensions.Caching.Distributed;
using Microsoft.Extensions.Caching.Redis;
using Microsoft.Extensions.Options;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace Redpoint.CommonFramework.Cache
{
    public class RetryableRedisCache : IDistributedCache, IDisposable
    {
        private readonly RedisCache _cache;
        private const int RetryCount = 10;

        public RetryableRedisCache(IOptions<RedisCacheOptions> redisOptions)
        {
            _cache = new RedisCache(redisOptions);
        }

        public void Dispose()
        {
            _cache.Dispose();
        }

        public byte[] Get(string key)
        {
            return Retry(() => _cache.Get(key));
        }

        public Task<byte[]> GetAsync(string key)
        {
            return RetryAsync(() => _cache.GetAsync(key));
        }

        public void Refresh(string key)
        {
            Retry(() => _cache.Refresh(key));
        }

        public Task RefreshAsync(string key)
        {
            return RetryAsync(() => _cache.RefreshAsync(key));
        }

        public void Remove(string key)
        {
            Retry(() => _cache.Remove(key));
        }

        public Task RemoveAsync(string key)
        {
            return RetryAsync(() => _cache.RemoveAsync(key));
        }

        public void Set(string key, byte[] value, DistributedCacheEntryOptions options)
        {
            Retry(() => _cache.Set(key, value, options));
        }

        public Task SetAsync(string key, byte[] value, DistributedCacheEntryOptions options)
        {
            return RetryAsync(() => _cache.SetAsync(key, value, options));
        }

        public Task<byte[]> GetAsync(string key, CancellationToken token = default(CancellationToken))
        {
            return RetryAsync(() => _cache.GetAsync(key, token));
        }

        public Task SetAsync(string key, byte[] value, DistributedCacheEntryOptions options, CancellationToken token = default(CancellationToken))
        {
            return RetryAsync(() => _cache.SetAsync(key, value, options, token));
        }

        public Task RefreshAsync(string key, CancellationToken token = default(CancellationToken))
        {
            return RetryAsync(() => _cache.RefreshAsync(key, token));
        }

        public Task RemoveAsync(string key, CancellationToken token = default(CancellationToken))
        {
            return RetryAsync(() => _cache.RemoveAsync(key, token));
        }

        #region Retry Logic

        private void Retry(Action func)
        {
            for (var i = 0; i < RetryCount; i++)
            {
                try
                {
                    func();
                    return;
                }
                catch (NullReferenceException ex)
                {
                    // Bug in Microsoft.Extensions.Caching.Redis! See https://github.com/aspnet/Caching/issues/270.
                    if (i == RetryCount - 1)
                    {
                        throw;
                    }

                    Thread.Sleep(i * 100);
                    continue;
                }
                catch (TimeoutException)
                {
                    if (i == RetryCount - 1)
                    {
                        throw;
                    }

                    Thread.Sleep(i * 100);
                    continue;
                }
            }

            throw new InvalidOperationException("Should not be able to escape for loop in Retry() of RetryableRedisCache");
        }

        private T Retry<T>(Func<T> func)
        {
            for (var i = 0; i < RetryCount; i++)
            {
                try
                {
                    return func();
                }
                catch (NullReferenceException ex)
                {
                    // Bug in Microsoft.Extensions.Caching.Redis! See https://github.com/aspnet/Caching/issues/270.
                    if (i == RetryCount - 1)
                    {
                        throw;
                    }

                    Thread.Sleep(i * 100);
                    continue;
                }
                catch (TimeoutException)
                {
                    if (i == RetryCount - 1)
                    {
                        throw;
                    }

                    Thread.Sleep(i * 100);
                    continue;
                }
            }

            throw new InvalidOperationException("Should not be able to escape for loop in Retry<T>() of RetryableRedisCache");
        }

        private async Task RetryAsync(Func<Task> func)
        {
            for (var i = 0; i < RetryCount; i++)
            {
                try
                {
                    await func();
                    return;
                }
                catch (NullReferenceException ex)
                {
                    // Bug in Microsoft.Extensions.Caching.Redis! See https://github.com/aspnet/Caching/issues/270.
                    if (i == RetryCount - 1)
                    {
                        throw;
                    }

                    await Task.Delay(i * 100);
                    continue;
                }
                catch (TimeoutException)
                {
                    if (i == RetryCount - 1)
                    {
                        throw;
                    }

                    await Task.Delay(i * 100);
                    continue;
                }
            }

            throw new InvalidOperationException("Should not be able to escape for loop in RetryAsync() of RetryableRedisCache");
        }

        private async Task<T> RetryAsync<T>(Func<Task<T>> func)
        {
            for (var i = 0; i < RetryCount; i++)
            {
                try
                {
                    return await func();
                }
                catch (NullReferenceException ex)
                {
                    // Bug in Microsoft.Extensions.Caching.Redis! See https://github.com/aspnet/Caching/issues/270.
                    if (i == RetryCount - 1)
                    {
                        throw;
                    }

                    await Task.Delay(i * 100);
                    continue;
                }
                catch (TimeoutException)
                {
                    if (i == RetryCount - 1)
                    {
                        throw;
                    }

                    await Task.Delay(i * 100);
                    continue;
                }
            }

            throw new InvalidOperationException("Should not be able to escape for loop in RetryAsync<T>() of RetryableRedisCache");
        }

        #endregion
    }
}
```

**Comment by jesuissur** [comment]
Hey

The problem seems to be on the double check assertion's side
I think the code in the Connect and ConnectAsync should be like that 

        private async Task ConnectAsync(CancellationToken token = default(CancellationToken))
        {
            token.ThrowIfCancellationRequested();

            if (_cache != null) // ===> and not _connection
            {
                return;
            }

            await _connectionLock.WaitAsync();
            try
            {
                if (_cache == null) // ===> again, not _connection
                {
                    _connection = await ConnectionMultiplexer.ConnectAsync(_options.Configuration);
                    _cache = _connection.GetDatabase();
                }
            }
            finally
            {
                _connectionLock.Release();
            }
        }

**Comment by riiight** [comment]
I believe we are also seeing this error sporadically. Unfortunately, I don't have a stack trace but was able to debug the server when it was occurring and was seeing the injected IDistributedCache instance was null.

**Comment by JunTaoLuo** [comment]
cc @muratg @Eilon The fixes in #378 is patch worthy, all the way back to 1.0.

**Comment by muratg** [comment]
@JunTaoLuo Feel free to file a bug for consideration! Thanks!
